File: server/api/grpc/namespace_test.go
/*
 * Copyright (C) 2017-2018 GIG Technology NV and Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package grpc

import (
	"context"
	"io/ioutil"
	"os"
	"path"
	"testing"

	"github.com/zero-os/0-stor/server"
	"github.com/zero-os/0-stor/server/api/grpc/rpctypes"
	pb "github.com/zero-os/0-stor/server/api/grpc/schema"
	"github.com/zero-os/0-stor/server/db"
	"github.com/zero-os/0-stor/server/db/badger"
	"github.com/zero-os/0-stor/server/encoding"

	"github.com/stretchr/testify/require"
)

func TestNewNamespaceAPIPanics(t *testing.T) {
	require.Panics(t, func() {
		NewNamespaceAPI(nil)
	}, "no db given")
}

func TestGetNamespace(t *testing.T) {
	require := require.New(t)

	api, clean := getTestNamespaceAPI(t)
	defer clean()

	data, err := encoding.EncodeNamespace(server.Namespace{Label: []byte(label)})
	require.NoError(err)
	err = api.db.Set(db.NamespaceKey([]byte(label)), data)
	require.NoError(err)

	req := &pb.GetNamespaceRequest{}

	resp, err := api.GetNamespace(context.Background(), req)
	require.Error(err)
	err = rpctypes.Error(err)
	require.Equal(rpctypes.ErrNilLabel, err)

	resp, err = api.GetNamespace(contextWithLabel(nil, label), req)
	require.NoError(err)

	require.Equal(label, resp.GetLabel())
	require.EqualValues(0, resp.GetNrObjects())
}

func getTestNamespaceAPI(t *testing.T) (*NamespaceAPI, func()) {
	require := require.New(t)

	tmpDir, err := ioutil.TempDir("", "0stortest")
	require.NoError(err)

	db, err := badger.New(path.Join(tmpDir, "data"), path.Join(tmpDir, "meta"))
	if err != nil {
		require.NoError(err)
	}

	clean := func() {
		db.Close()
		os.RemoveAll(tmpDir)
	}

	require.NoError(err)
	api := NewNamespaceAPI(db)

	return api, clean
}

File: daemon/client/py-client/pydaemon/namespace_test.py
# Copyright (C) 2017-2018 GIG Technology NV and Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest
from unittest import mock

from .generated import daemon_pb2_grpc as stubs
from .generated import daemon_pb2 as model

from .namespace import Namespace


class TestNamespaceClient(unittest.TestCase):
    def setUp(self):
        with mock.patch.object(stubs, 'NamespaceServiceStub') as m:
            m.side_effect = mock.MagicMock()
            self.client = Namespace(None)

    def test_created(self):
        self.assertIsNotNone(self.client)
        self.assertIsInstance(self.client._stub, mock.MagicMock)

    def test_list(self):
        self.client.list()

        self.client._stub.ListNamespaces.assert_called_once_with(
            model.ListNamespacesRequest()
        )

    def test_create(self):
        namespace = 'namespace'

        self.client.create(namespace)

        self.client._stub.CreateNamespace.assert_called_once_with(
            model.CreateNamespaceRequest(
                namespace=namespace
            )
        )

    def test_delete(self):
        namespace = 'namespace'

        self.client.delete(namespace)

        self.client._stub.DeleteNamespace.assert_called_once_with(
            model.DeleteNamespaceRequest(
                namespace=namespace
            )
        )

    def test_get_permission(self):
        namespace = 'namespace'
        user = 'test'

        self.client.get_permission(namespace, user)

        self.client._stub.GetPermission.assert_called_once_with(
            model.GetPermissionRequest(
                namespace=namespace,
                userID=user
            )
        )

    def test_set_permission(self):
        namespace, user, admin, read, write, delete =\
            'namespace', 'test', True, False, True, False

        self.client.set_permission(namespace, user, admin, read, write, delete)

        self.client._stub.SetPermission.assert_called_once_with(
            model.SetPermissionRequest(
                namespace=namespace,
                userID=user,
                permission=model.Permission(
                    admin=admin, read=read, write=write, delete=delete,
                )
            )
        )
